core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java

```diff
@@ -322,6 +327,24 @@ private SdkClientConfiguration finalizeConfiguration(SdkClientConfiguration conf
                      .build();
     }
 
+    /**
+     * Add an IdentityProvider to the IdentityProviders configured for the client.
+     */
+    protected IdentityProviderConfiguration updateIdentityProviderConfiguration(SdkClientConfiguration config,
+                                                                                IdentityProvider<?> identityProvider) {
+        // Currently it is not possible for identityProvider to be null as default provider is used while building the client if
+        // the clientConfig is null. However, we do want to support ability to unset a identity provider later.
+        // Moreover, putIdentityProvider will throw NPE on null, so adding the null check here. Also, validateClientOptions
+        // currently asserts it is not null, which will have to change when we allow unsetting default identity provider.
+        if (identityProvider == null) {
+            return config.option(IDENTITY_PROVIDER_CONFIGURATION);
+        }
+        return config.option(IDENTITY_PROVIDER_CONFIGURATION)
+                     .toBuilder()
+                     .putIdentityProvider(identityProvider)
+                     .build();
+    }
```

Comment:
This cross-module API feels like it's going to be a pain to maintain, long-term. Can we just do the null check where we make the call, instead?

Reply:
Move the null check to call site in https://github.com/aws/aws-sdk-java-v2/pull/4111/commits/5b0652e8da4d6fbc4abff7a150483b7568d2832c

Reply:
Moved the full logic to call site in https://github.com/aws/aws-sdk-java-v2/pull/4111/commits/5ae9bff1853b993474f2b48e0e9d435a94b7326a